compiler/system.c

```diff
@@ -53,23 +53,51 @@
 #include "proto.h"
 
 #ifdef __ELKS__
-void *fmemrealloc(void *q, size_t size)
+
+#define MAX_NEAR_ALLOC  64U   /* max size to allocate from near heap */
+
+#define SEGMENT(ptr)    ((unsigned long)(char __far *)(ptr) >> 16)
+
+void __far *memalloc(unsigned long size)
 {
-    void *p;
-
-    if (!q)
-        return fmemalloc(size);
-    p = fmemalloc(size);
-    if (p) {                    /* on fail, previous memory not freed */
-        memcpy(p, q, size);     /* FIXME copies too much!! */
-        fmemfree(q);
-    }
+	char *p;
+	char __far *fp;
+	if (size <= MAX_NEAR_ALLOC)
+	{
+		p = malloc((unsigned int)size);
+		fp = (void __far *)p;
+	}
+	else
+	{
+		fp = fmemalloc(size);
+	}
+	return fp;
+}
 
-    if (!p)
-        printf("Error in realloc\n");
-    return p;
+void memfree(void __far *ptr)
+{
+	if (ptr == 0)
+		return;
+	if (SEGMENT(ptr) == SEGMENT(&ptr)) /* near pointer */
+		free((char *)ptr);
+	else
+		fmemfree(ptr);
+}
 
+void __far *memrealloc(void __far *ptr, unsigned long size)
+{
+	void __far *new;
+	if (!ptr)
+		return memalloc(size);
+	else
+	{
+		new = memalloc(size);
+		memcpy(new, ptr, size);    /* FIXME copies too much!! */
+		memfree(ptr);
+	}
+	return new;
 }
+
 #endif
```

Comment:
See comment above - how about the idea of moving all this code out of system.c into mem.c? The same exact mem.c that would be used in all the tools. This is the only way we'll know we're ready to remove the memory functions into the C library, because they'd all be exactly the same - in mem.c. This idea also keeps all our special code out of the tool's .c files, providing better separation.

Reply:
Yes, but first we need to improve the api for setting the MAX_NEAR_ALLOC, for example.

Reply:
> first we need to improve the api for setting the MAX_NEAR_ALLOC

Agreed, I didn't quite realize you've not been able to use the same value of MAX_NEAR_ALLOC. But each copy of mem.c could have a different value for MAX_NEAR_ALLOC, with the other portions the same. This would still keep our memory allocation routines out of the tool main source code, for example.